onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -2073,22 +2074,20 @@ static Status RegisterCudaKernels(KernelRegistry& kernel_registry) {
 
 }  // namespace cuda
 
-static std::shared_ptr<onnxruntime::KernelRegistry> s_kernel_registry;
+static std::shared_ptr<KernelRegistry> CreateCudaKernelRegistry() {
+  std::shared_ptr<KernelRegistry> registry = KernelRegistry::Create();
+  ORT_THROW_IF_ERROR(cuda::RegisterCudaKernels(*registry));
+  return registry;
+}
+
+static SharedPtrThreadSafeWrapper<KernelRegistry> s_kernel_registry{&CreateCudaKernelRegistry};
```

Comment:
I would suggest not having global initializers. If you change it to function local, things will be much simpler.

Reply:
Function local static would be nice, but it needs to be reset by another function. @RyanUnderhill knows more about why.

Reply:
Even it is function local, you still can call Reset() manually. Function local delays initiation to run time.

Reply:
I guess you mean doing something like:
```c++
static std::shared_ptr<KernelRegistry>& CudaKernelRegistry() {
  static std::shared_ptr<KernelRegistry> registry = ...
  return registry;
}

std::shared_ptr<KernelRegistry> CUDAExecutionProvider::GetKernelRegistry() const {
  return CudaKernelRegistry();
}

void Shutdown_DeleteRegistry() {
  CudaKernelRegistry().reset();
}
```

But that has different behavior when calls to GetKernelRegistry() and Shutdown_DeleteRegistry() are interleaved. E.g., GetKernelRegistry() will return an empty shared_ptr after Shutdown_DeleteRegistry() is called.

Reply:
I think we don't really need to ref count it. Shutdown_DeleteRegistry should only be called once when the DLL is unloading. And, nobody should call GetKernelRegistry() again. Nobody would call Shutdown_DeleteRegistry  manually. It is an internal function that only serve the needs of Ryan's provider bridge code.

Reply:
In the Java tests, apparently GetKernelRegistry() gets called after Shutdown_DeleteRegistry() and that causes a test failure. So I'll go back to the original approach.

Reply:
Hi @Craigacp, the Java API seems to support creating and destroying an ORT environment multiple times within a process (e.g., in unit tests), but according to some other folks on the ORT team, we only expect one environment to be created and destroyed once in actual scenarios. Some of the previous comments in this thread have this assumption.

I wanted to understand why the Java API works this way and if it is indeed necessary to create and destroy an ORT environment multiple times within one process. For comparison, in the C# API, the environment class is a singleton without the explicit reference count.
https://github.com/microsoft/onnxruntime/blob/d52d3c00520a10075e3bc75fe90db5b4c83ee557/csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs#L71-L76

Reply:
It works that way because while I understood there was supposed to only be at most one environment at any given time, I didn't realise that there could only be a single one in any given process. Given there are user controlled options which need to be set on environment construction I can't automatically create a singleton instance in the static initializer block for the `OrtEnvironment` class, as that wouldn't allow users to setup things like thread pools or logging, which is why we ended up with the current approach. I can refactor the code so the environment isn't closeable and rely on a shutdown hook to close it, but then there's no way for a user to signal that they are finished with ORT and it can release it's resources. The shutdown hook isn't guaranteed to execute before the JVM goes away, but I don't know if that matters much.

How does the C# code shutdown the environment? Is it done when the runtime terminates or can the user signal this? Also, how does it deal with the environment construction options like thread pools and logging?

Reply:
> It works that way because while I understood there was supposed to only be at most one environment at any given time, I didn't realise that there could only be a single one in any given process until you said it just now. Given there are user controlled options which need to be set on environment construction I can't automatically create a singleton instance in the static initializer block for the `OrtEnvironment` class, as that wouldn't allow users to setup things like thread pools or logging, which is why we ended up with the current approach. I can refactor the code so the environment isn't closeable and rely on a shutdown hook to close it, but then there's no way for a user to signal that they are finished with ORT and it can release its resources. The shutdown hook isn't guaranteed to execute before the JVM goes away, but I don't know if that matters much.

@pranavsharma What do you think?

> How does the C# code shutdown the environment? Is it done when the runtime terminates or can the user signal this? Also, how does it deal with the environment construction options like thread pools and logging?

I think the C# ORT environment will be shutdown when the runtime terminates. The creation/shutdown is tied to a lazily initialized, static class variable.

https://github.com/microsoft/onnxruntime/blob/d52d3c00520a10075e3bc75fe90db5b4c83ee557/csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs#L51
https://github.com/microsoft/onnxruntime/blob/d52d3c00520a10075e3bc75fe90db5b4c83ee557/csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs#L57
https://github.com/microsoft/onnxruntime/blob/d52d3c00520a10075e3bc75fe90db5b4c83ee557/csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs#L175

As for environment configuration, that looks to be unsupported (currently hardcoded).
https://github.com/microsoft/onnxruntime/blob/d52d3c00520a10075e3bc75fe90db5b4c83ee557/csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs#L54

Reply:
@Craigacp We can match the same behavior as C# for Java as well. Which part won't be a singleton like C#? C# does _not_ allow the user to configure it on instantiation.

Reply:
I'd prefer not to remove useful functionality like configuring a threadpool and setting the logging level. Switching it over to a singleton will mean the environment is constructed at an unpredictable time (as it's based on class initialization) and there's no potential for a user to close it to free up resources while keeping the JVM alive. A shutdown hook will clean up the environment most of the time (apart from when the JVM gets SIGKILLed or similar) and we'd need that for the singleton anyway. Allowing user controlled construction avoids the first problem while giving more flexibility.

What kind of things live inside the environment object? Is it worth allowing users to close it explicitly if they are finished with ORT, under the assumption that it throws some kind of exception if they try to make another one in the same process?

Reply:
I've filed #10670 which switches the environment over so that only one can be created in the lifetime of the JVM (unless the user starts messing with class loaders), and its closed by a JVM shutdown hook. I left the close method in place for compatibility though it is now a no-op. If we want the environment to be closeable to release resources then I'll need to do some more refactoring as the current close idiom will be very confusing.

Reply:
In terms of resources the most important one that the Environment contains is the threadpool and that too only if you create the env with the global threadpool option. See [here](https://github.com/microsoft/onnxruntime/blob/4e9e01cb3c008335a2471c27dbdf7dd5d12e4224/include/onnxruntime/core/session/environment.h#L90-L94). Besides this there are global registrations for various schemas that can't be undone.

The OrtEnv obj exposed in the C API wraps the internal Environment class. Each time you request an OrtEnv object the refcount is [incremented for the same instance](https://github.com/microsoft/onnxruntime/blob/4e9e01cb3c008335a2471c27dbdf7dd5d12e4224/onnxruntime/core/session/ort_env.cc#L81) in the process. We do provide a [Release method](https://github.com/microsoft/onnxruntime/blob/4e9e01cb3c008335a2471c27dbdf7dd5d12e4224/onnxruntime/core/session/ort_env.cc#L85-L95) that decrements the refcount and finally gets rid of the instance allowing another instance to be created in the same process. The OrtEnv destructor [unloads the shared libs](https://github.com/microsoft/onnxruntime/blob/4e9e01cb3c008335a2471c27dbdf7dd5d12e4224/onnxruntime/core/session/ort_env.cc#L38-L43).

Reply:
I used to have a similar refcount in the Java side and it would hand out the same Java reference to the native OrtEnv, but it would call the destructor when the count gets to zero and allow users to build a fresh environment after that, which is the source of the trouble. It doesn't look like there are too many resources held by it, so it's probably ok without an explicit close.